GET POST AND DELETE Reminder API Created

alian1001 opened the pull request:
EMAIL IS USED FOR
-Reminder update, add, delete
-Reminder when time is triggered
-This clinician has invited you to connect, if user not registered, send invitation token, otherwise add to contacts book


POST

{
  "setById": "user_2kb0nnXRX4SdwJAO5gVMyF12Xmw",
  "setForId": "user_2kit0iBGwimRaKgg3WjIIqZPAvG",
  "scheduledAt": "2024-08-20T10:00:00Z",
  "startDate": "2024-08-19T00:00:00Z",
  "endDate": "2024-09-19T00:00:00Z",
  "description": "Appointment reminder",
  "type": "APPOINTMENT",
  "intervalInDays": 30
}


PUT

{
  "id": 1,
  "setById": "user_2kb0nnXRX4SdwJAO5gVMyF12Xmw",
  "setForId": "user_2kit0iBGwimRaKgg3WjIIqZPAvG",
  "scheduledAt": "2024-08-21T10:00:00Z",
  "startDate": "2024-08-19T00:00:00Z",
  "endDate": "2024-09-20T00:00:00Z",
  "description": "Updated appointment reminder",
  "type": "APPOINTMENT",
  "intervalInDays": 60,
  "isReminderActive": true
}

DELETE

{
  "id": 1
}


Do let me know if anything else has gone wrong.

alian1001 committed:
Email Template

alian1001 committed:
package-update

alian1001 committed:
fix

alian1001 committed:
fix

alian1001 committed:
Added Reminder API

Reminder API added, POST, PUT, and DELETE

alian1001 committed:
send email function

alian1001 committed:
Update route.ts

alian1001 committed:
cleanup

ellenxxxiao commented on src/app/api/reminder/route.ts:
```diff
@@ -0,0 +1,241 @@
+import { EmailTemplate } from "../../components/emailTemplate/email-template";
+
+const prisma = new PrismaClient();
+const resend = new Resend(process.env.RESEND_API_KEY);
```
Update the env template `.env.example` to include this key.

ellenxxxiao commented on src/app/api/reminder/route.ts:
```diff
@@ -0,0 +1,241 @@
+  try {
+    const { data, error } = await resend.emails.send({
+      from: "OrthoChat <onboarding@resend.dev>",
+      to: ["alian.haidar01@gmail.com"], // Hardcoded email recipient
```
accept recipient email as param

alian1001 replied:
no domain to be used. stick with hardcoded for now.

ellenxxxiao commented on src/app/api/email/route.ts:
Is this a test route? If yes you can remove it

ellenxxxiao commented on src/app/login/page.tsx:
Same here, remove code for testing purposes when submitting PR

ellenxxxiao commented on src/app/components/emailTemplate/email-template.tsx:
Try using Tailwind CSS classes.
Could also make use of React Email to use some components. Refer to https://resend.com/docs/dashboard/emails/email-templates

ellenxxxiao commented on src/app/api/reminder/route.ts:
```diff
@@ -0,0 +1,241 @@
+}
+
+// PUT: Update an existing reminder
+export async function PUT(request: Request) {
```
Use `PATCH`

ellenxxxiao commented on src/app/api/reminder/route.ts:
```diff
@@ -0,0 +1,241 @@
+// GET: Fetch reminders
+export async function GET(request: Request) {
+  try {
+    const url = new URL(request.url);
+    const setById = url.searchParams.get("setById");
+    const setForId = url.searchParams.get("setForId");
+    const isReminderActive =
+      url.searchParams.get("isReminderActive") === "true";
```
Could be simplified by:

```typescript
import { NextRequest, NextResponse } from 'next/server';

export async function GET(
  request: NextRequest,
) {
  //...

  // get search params
  const searchParams = request.nextUrl.searchParams;
  const isActive = searchParams.get("isActive");

  //...
}

```

Refer to https://nextjs.org/docs/app/api-reference/functions/next-request#nexturl

ellenxxxiao commented on src/app/api/reminder/route.ts:
```diff
@@ -0,0 +1,241 @@
+      await request.json();
+
+    if (!id) {
+      return new Response(
+        JSON.stringify({ error: "Reminder ID is required" }),
+        {
+          status: HttpStatusCode.BadRequest,
+          headers: { "Content-Type": "application/json" }
+        }
+      );
+    }
+
+    await prisma.reminder.delete({
+      where: { id }
+    });
+
+    // Send email when a reminder is deleted
+    await sendEmail("Reminder Deleted", "deleted", {
+      firstName: "Patient",
+      description: `Your Reminder has been deleted successfully.`,
+      startDate: startDate,
+      endDate: endDate,
+      type,
+      intervalInDays
+    });
+
+    return new Response(
+      JSON.stringify({ message: "Reminder deleted successfully" }),
+      {
+        status: HttpStatusCode.Ok,
+        headers: { "Content-Type": "application/json" }
+      }
+    );
+  } catch (error) {
+    return new Response(JSON.stringify({ error: getErrorMessage(error) }), {
+      status: HttpStatusCode.InternalServerError,
+      headers: { "Content-Type": "application/json" }
+    });
+  }
+}
```
`PATCH` and `DELETE` operations should be performed on a specific reminder only, so these two endpoints should in file `api/reminder/[id]/route.ts`. The param `id` refers to the reminder ID. Refer to the `/users` endpoint.

ellenxxxiao commented on src/app/api/reminder/route.ts:
```diff
@@ -0,0 +1,241 @@
+export async function POST(request: Request) {
+  try {
+    const {
+      setById,
+      setForId,
+      scheduledAt,
+      startDate,
+      endDate,
+      description,
+      type,
+      intervalInDays
```
Could be converted into a type

ellenxxxiao reviewed (changes requested):
Nice work, refer to comments

alian1001 committed:
Merge branch 'main' of https://github.com/ellenxxxiao/uwa-orthodontics-app into issue-38-Email_Notification

alian1001 committed:
package fix

ellenxxxiao committed:
removed login

ellenxxxiao committed:
seed reminders table; updated prisma client

ellenxxxiao committed:
typo

ellenxxxiao committed:
get, patch, del by id

ellenxxxiao committed:
update & migrate

ellenxxxiao committed:
post in progress

ellenxxxiao committed:
typo fix

ellenxxxiao committed:
updated post

ellenxxxiao committed:
rmved test route

ellenxxxiao committed:
linting

alian1001 committed:
Cleanup

alian1001 committed:
Fix reminder

alian1001 committed:
Fix API call

Fix route. Now it pulls all reminders from the user logged on. and DELETE and PATCH API's now have email functionality.

alian1001 committed:
up fix

Now API can send email, and reads reminders from user logged in only.

alian1001 committed:
Cleanup code

Passed checks

alian1001 committed:
Merge branch 'main' of https://github.com/ellenxxxiao/uwa-orthodontics-app into issue-38-Email_Notification

alian1001 committed:
jsonfix

ellenxxxiao committed:
endpoints

ellenxxxiao committed:
commented out email

ellenxxxiao committed:
filter by setBy current logged in user

ellenxxxiao committed:
fetch reminders from endpoint

ellenxxxiao committed:
formatted reminders in API response

ellenxxxiao committed:
populate user dropdown with actual users;
functional edit reminders

ellenxxxiao committed:
button to delete reminder; toast to indicate update success

ellenxxxiao committed:
add new reminder

ellenxxxiao committed:
truncate longer names on cards

ellenxxxiao committed:
removed settings nav item

ellenxxxiao committed:
lint
